src/internal/internal_geset.cc

```diff
@@ -92,112 +92,153 @@ void set(internal::TargetType<Target::Devices>,
-                    if (A.tileIsLocal(i, j) && device == A.tileDevice(i, j)) {
-                        A_tiles_set.insert({i, j});
+                    if (A.tileIsLocal( i, j ) && device == A.tileDevice( i, j )) {
+                        A_tiles_set.insert( { i, j } );
                     }
                 }
             }
-            A.tileGetForWriting(A_tiles_set, device, LayoutConvert(layout));
+            A.tileGetForWriting( A_tiles_set, device, layout );
+
+            scalar_t** a_array_host = A.array_host( device, queue_index );
 
-            scalar_t** a_array_host = A.array_host(device, queue_index);
+            // If offdiag == diag value, lump diag tiles with offdiag tiles
+            // in one batch.
+            bool diag_same = offdiag_value == diag_value;
 
+            // Build batch groups.
             int64_t batch_count = 0;
-            int64_t mb[8], nb[8], lda[8], group_count[8];
-            for (int q = 0; q < 4; ++q) {
-                group_count[q] = 0;
-                lda[q] = 0;
-                mb[q] = A.tileMb(irange[q][0]);
-                nb[q] = A.tileNb(jrange[q][0]);
-                for (int64_t i = irange[q][0]; i < irange[q][1]; ++i) {
-                    for (int64_t j = jrange[q][0]; j < jrange[q][1]; ++j) {
-                        if (A.tileIsLocal(i, j) && device == A.tileDevice(i, j)) {
-                            if (i != j) {
-                                a_array_host[batch_count] = A(i, j, device).data();
-                                lda[q] = A(i, j, device).stride();
-                                ++group_count[q];
-                                ++batch_count;
-                            }
+            struct Params {
+                int64_t count, mb, nb, lda;
+                scalar_t diag_value;
+            };
+            std::vector<Params> group_params;
+            for (size_t jj = 0; jj < jrange.size() - 1; ++jj) {
+            for (size_t ii = 0; ii < irange.size() - 1; ++ii) {
+                Params group = { 0, -1, -1, -1, offdiag_value };
+                for (int64_t j = jrange[ jj ]; j < jrange[ jj+1 ]; ++j) {
+                for (int64_t i = irange[ ii ]; i < irange[ ii+1 ]; ++i) {
+                    if ((diag_same || i != j)
+                        && A.tileIsLocal( i, j )
+                        && device == A.tileDevice( i, j ))
+                    {
+                        auto Aij = A( i, j, device );
+                        a_array_host[ batch_count ] = Aij.data();
+                        if (group.count == 0) {
+                            group.mb  = Aij.mb();
+                            group.nb  = Aij.nb();
+                            group.lda = Aij.stride();
+                        }
+                        else {
+                            assert( group.mb  == Aij.mb() );
+                            assert( group.nb  == Aij.nb() );
+                            assert( group.lda == Aij.stride() );
                         }
```

Comment:
Is this piece of code needed? The way I understand it is more we do not trust enough the `irange` and `jrage`, correct?

Reply:
The checks on `mb` and `nb` are probably unnecessary, yea.  I'll look at removing them.  The check on `lda` could come into play if users are providing memory themselves and have inconsistent strides.  It's maybe not a case that will actually arise, but assertions are nicer to debug if it does.

Reply:
They're asserts to ensure the program logic is correct. If compiled with -DNDEBUG, they disappear. I would leave them, though maybe they don't need to go into every function that gets regions.